fuel-core/src/executor.rs

```diff
@@ -897,67 +903,175 @@ impl Executor {
-                    Transaction::Mint(_) => return Ok(()),
-                };
+            ExecutionTypes::Validation(tx) => {
+                for input in tx.inputs() {
+                    match input {
+                        Input::CoinSigned { tx_pointer, .. }
+                        | Input::CoinPredicate { tx_pointer, .. } => {
+                            let _ = tx_pointer;
+                            // TODO: Also calculate `tx_pointer` based on `utxo_id` pointer.
+                            // if tx_pointer != &TxPointer::new(height.into(), idx) {
+                            //     return Err(Error::InvalidTransactionOutcome {
+                            //         transaction_id: tx.id(),
+                            //     })
+                            // }
+                        }
+                        Input::Contract {
+                            utxo_id,
+                            balance_root,
+                            state_root,
+                            contract_id,
+                            ..
+                        } => {
+                            let mut contract = ContractRef::new(&mut *db, contract_id);
+                            if *utxo_id != expected_utxo_id(&contract)? {
+                                return Err(Error::InvalidTransactionOutcome {
+                                    transaction_id: tx.id(),
+                                })
+                            }
+                            if balance_root != &contract.balance_root()? {
+                                return Err(Error::InvalidTransactionOutcome {
+                                    transaction_id: tx.id(),
+                                })
+                            }
+                            if state_root != &contract.state_root()? {
+                                return Err(Error::InvalidTransactionOutcome {
+                                    transaction_id: tx.id(),
+                                })
+                            }
+                            // TODO: Also check `tx_pointer` based on `utxo_id` pointer.
+                        }
+                        _ => {}
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
 
-                let iter = inputs.iter().filter_map(|input| match input {
-                    Input::Contract {
-                        utxo_id,
-                        contract_id,
-                        ..
-                    } => Some((utxo_id, contract_id)),
-                    _ => None,
-                });
-                for (utxo_id, contract_id) in iter {
-                    if *utxo_id != expected_utxo_id(contract_id)? {
-                        return Err(Error::InvalidTransactionOutcome {
-                            transaction_id: tx.id(),
-                        })
+    // TODO: Maybe we need move it to `fuel-vm`? O_o Because other `Outputs` are processed there
```

Comment:
I'm unsure about this part. `fuel-vm` already updates outputs for all utxos. Maybe better to move it there or extract all updates here.

Reply:
fuel-vm could set the balance and state roots, but it doesn't manage utxo id's or utxo state directly. This is mainly to enable the client to do bulk optimization for utxo lookups and validation across many txs at once in parallel.